### DISCUSSION TITLE
Add method for packageRule to filter on environment var

### DISCUSSION THREAD

**Discussion by namrog84** [renovatebot/renovate#12729] [Suggest an Idea]
We have a large mono repo.
e.g. we have 20 different projects. 
only 1 of them needs renovateBot to run on a mac,  the other 19 can run on ubuntu or windows build agents. 
Ideally we'd prefer to run on ubuntu/windows generally. 

It'd be great if there was some kind of built in mechanic to filter or do this.  Allow Renovate to only handle certain packages from 1 build agent,  and another build agent to handle other packages.  

One possible idea:

Since we are selfhosted, and using packageRules it'd be great if we could configure 2 build agents that run RenovateBot (e.g. ubuntu runs at 9am, mac runs at 9pm), that had different environment variables set before running `npx renovate`, Such environment platforms such as "platform=mac" and "platform=other"  (key=value)

Is there any way to filter on like RENOVATE_USER_AGENT or something from the packageRules?

I do see a thing for detectHostRulesFromEnv, perhaps there is a way to leverage this? But it looks like its only for token, username and password and not for packageRules?

Then in the packageRules in our repo.

```
{
      // disable all packages, and selectively enable them because of order of rules overriding earlier defined rules.
      ...
      "enabled": false
},
{
      "matchEnvironmentVar": {  // new feature will compare against environment variables on machine.
            "platform": "other"
       },
      "matchPaths": ["src/web/randomReactpackage/package.json"],,
      "addLabels": ["beingRunByWhateverMachine"],
      "enabled": true
    },
{
      "matchEnvironmentVar": { // new feature will compare against environment variables on machine.
            "platform": "mac"
       },
      "matchPaths": ["src/apps/RNApp/package.json"],
      "addLabels": ["beingRunFromMacRenovate"]
    },      
    "enabled": true
```

We could use this matchEnvironmentVar to only run this rule if it found a matching key/value environment, then we could put our podInstall or whatever in that one.  Otherwise anything that didn't have matching wouldn't run that one. 

Idea 2.

It'd be most amazing if renovate.json could instead be renovate.js and we could just slap some process.env.whateverVar for like `"enabled": process.env.externalBool`

**Comment by rarkins** [comment]
Global config allows you to configure with `.js` today

**Reply by namrog84** [reply]
Ah super awesome. I'll investigate trying that.
I had thought the config.js was strictly for things like auth and which repos. 

I didn't know I can put packageRules and all the same things as renovate.json can take? 


Does that mean I don't need a renovate.json in the repo if I have everything properly configured for the repo in the dependabot repo's config.js?

**Reply by namrog84** [reply]
I just found the statement in the doc 
"Your bot's global config can include both global as well as non-global configuration options, while user/repo config can only include non-global options."

I had no idea.  mind blown!

This opens up so many possibilities!

**Reply by viceice** [reply]
https://docs.renovatebot.com/self-hosted-configuration/#requireconfig

**Reply by namrog84** [reply]
Dang.  Mind blown again. I was just wondering how I was going to do that with just a barely empty renovate.json or something.  You read my mind!  Y'all are just the best

**Reply by rarkins** [reply]
We recommend that you generally put as much of the repo config in the repo as you can, but yes you can put everything in the global config instead if that works for (even skipping having renovate.json files in the repo altogether if you really want - see the `requireConfig` option)

**Reply by namrog84** [reply]
I take it the local repo's renovate.json will override anything set in the base global config?   Or is global take precendence.

Assuming there were 2 conflicting rules or settings.

I'll definitely play around with it a bit and do as much in the repo as I can, but it definitely opens up some possibilities for some scenarios I had been struggling with recently.

**Reply by rarkins** [reply]
Repo config takes precedence

**Reply by namrog84** [reply]
Excellent, I will plan accordingly then!
